src/main/java/com/rackspace/jenkins_nodepool/Janitor.java

```diff
@@ -160,40 +181,104 @@ void cleanNode(NodePoolSlave nodePoolSlave, String reason) {
     boolean hasInvalidLabel(NodePoolSlave nodePoolSlave) {
         final String nodeLabel = nodePoolSlave.getLabelString();
         final List<NodePool> nodePools = NodePools.get().nodePoolsForLabel(new LabelAtom(nodeLabel));
-        return (nodePools == null || nodePools.size() == 0);
+        return (nodePools == null || nodePools.isEmpty());
     }
 
     /**
-     * Check if the given slave node has disappeared in ZooKeeper
-     *
-     *
-     * @param nodePoolSlave the node to check
-     * @return true if the node has disappeared
+     * A routine to review and release any stale locks that Jenkins may have.
      */
-    private boolean isMissing(NodePoolSlave nodePoolSlave) {
-        NodePoolComputer c = (NodePoolComputer) nodePoolSlave.toComputer();
-        if (c == null || c.isOffline()) {
-            // agent is offline - confirm that this node still exists in ZK/NP:
-            final NodePoolNode nodePoolNode = nodePoolSlave.getNodePoolNode();
+    private void cleanStaleLocks() {
+        LOG.log(FINE, "--------------------- Janitor scanning -------------------");
 
-            try {
-                if (nodePoolNode == null) {
-                    // node is unknown
-                    LOG.log(Level.FINE, "Slave " + nodePoolSlave + " has no associated Node record.");
-                    return true;
-                }
-                if (!nodePoolNode.exists()) {
-                    // Corresponding ZNode is gone, this is definitely an orphan record in Jenkins
-                    LOG.log(Level.FINE, "Slave " + nodePoolSlave + " no longer exists in ZK.");
-                    return true;
+        // Only interested in NP nodes
+        if (NodePools.get().getNodePools().isEmpty()) {
+            LOG.log(WARNING, "Empty NodePool list - unable to query for requests and nodes.");
+            return;
+        }
+
+        try {
+            LOG.log(FINE, "Looking for nodes...");
+            final String zkNodesRootPath = format("/%s", NodePools.get().getNodePools().get(0).getNodeRoot());
+            final List<String> nodesPaths = conn.getChildren().forPath(zkNodesRootPath);
+            if (nodesPaths.isEmpty()) {
+                LOG.log(FINE, "No NodePool nodes registered. Expecting reserved nodes (min-ready).");
+            } else {
+                for (String nodePath : nodesPaths) {
+                    // Read and parse the data into a POJO model
+                    final String path = format("%s/%s", zkNodesRootPath, nodePath);
+                    final byte[] data = conn.getData().forPath(path);
+                    final ZKNodeModel model = gson.fromJson(new String(data, StandardCharsets.UTF_8), ZKNodeModel.class);
```

Comment:
I think ZKNodeModel should integrate with NodePoolNode somehow, seems unnecessary to have two classes representing the same thing?